### PR TITLE
Persistent analysis, part I

### LINKED ISSUE #1112: Proxy types backed by index database
To be used when declaring imported items when persistent model is available or when analysis is still running and we used Python informational process.

### LINKED ISSUE #1177: API to determine unique package id
Need a way to unique identify packages, such `name + version + ?` perhaps. If it is a version, then we needs means to reliably determine version of a package. If package is not versioned, perhaps hash the content?

### LINKED ISSUE #1178: Analysis DB schema (model)
Classes for the persistent analysis model.

### LINKED ISSUE #1205: Basic restore saved analysis from db
Initial test to save/restore analysis of builtins module

### LINKED ISSUE #1221: Tell between class/member chain vs module import
Specifically, handle `a.b.c` where `a` is a class with inner class `b` and method `c` vs. case where `a.b.c` is an package and `b` will not appear in members of `b` via `GetMemberNames`

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Closes #1112 
Closes #1177 
Closes #1178 
Closes #1205 
Closes #1221

Works in basic scenarios like `import sys`. Should be merged in inactive state - remove creation of the database service in Server initialization by commenting out

```cs
_services.AddService(new ModuleDatabase(_services));
```

- Introduces 'qualified name' to types. This is how type is identified in the module saved analysis. For example, `A.B.C` for method `C` inside method `B` in class `A`. `i:mod:z` is an instance of type `z` defined in module `mod`.
- Provides model classes for modules, classes, functions, properties, variables.
- Fixes issues when declaring module for `Union` and collections from `typing` was actually `builtins`.
- Uses LiteDb to save models to file. Currently uses one file per module for convenience of debugging.
- Splits `PythonModule` to partial classes.
- Provides code to recostruct types from qualified names.
- Implements database service to get module constructed from the db instead of running analysis.

See outstanding issues in the epic on what is not handled yet.

### CODE REVIEW

File: src/Analysis/Ast/Impl/Types/Collections/PythonCollectionType.cs [hunk 1]
```diff
@@ -32,17 +32,17 @@ internal class PythonCollectionType : PythonTypeWrapper, IPythonCollectionType {
...
+        /// <param name="declaringModule">Declaring module.</param>
         /// <param name="isMutable">Indicates if collection is mutable (like list) or immutable (like tuple).</param>
         public PythonCollectionType(
             string typeName,
             BuiltinTypeId collectionTypeId,
-            IPythonInterpreter interpreter,
+            IPythonModule declaringModule,
             bool isMutable
-        ) : base(collectionTypeId, interpreter.ModuleResolution.BuiltinsModule) {
+            ) : base(collectionTypeId, declaringModule) {
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Types/Collections/PythonCollectionType.cs]: Is this change related to databases or a separate bug fix?

File: src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs [hunk 2]
```diff
@@ -28,10 +28,12 @@ internal sealed class TypingType : LocatedMember, IPythonType {
         public TypingType(IPythonModule declaringModule, IPythonType type): base(declaringModule) {
             _type = type ?? throw new ArgumentNullException(nameof(type));
             Name = $"Type[{_type.Name}]";
+            QualifiedName = $"{DeclaringModule.Name}:Type[{_type.QualifiedName}]";
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs]: Can `DeclaringModule` be null or have no name?

File: src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Collections.cs [hunk 3]
```diff
@@ -110,14 +110,14 @@ internal sealed partial class ExpressionEval {
...
+                        return PythonCollectionType.CreateList(Module, new[] { v1 });
                     case SetComprehension sc:
                         var v2 = GetValueFromExpression(sc.Item) ?? UnknownType;
-                        return PythonCollectionType.CreateSet(Interpreter, new[] { v2 });
+                        return PythonCollectionType.CreateSet(Module, new[] { v2 });
                     case DictionaryComprehension dc:
                         var k = GetValueFromExpression(dc.Key) ?? UnknownType;
                         var v = GetValueFromExpression(dc.Value) ?? UnknownType;
-                        return new PythonDictionary(new PythonDictionaryType(Interpreter), new Dictionary<IMember, IMember> { { k, v } });
+                        return new PythonDictionary(new PythonDictionaryType(Interpreter.ModuleResolution.BuiltinsModule), new Dictionary<IMember, IMember> { { k, v } });
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Collections.cs]: Why is it `BuiltinsModule` here and not the `Module`?

File: src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs [hunk 4]
```diff
@@ -258,6 +260,10 @@ public ModuleWalker(IServiceContainer services, IPythonModule module)
 
                         var memberType = member?.GetPythonType();
                         var stubMemberType = stubMember.GetPythonType();
+
+                        if (builtins.Equals(memberType?.DeclaringModule) || builtins.Equals(stubMemberType.DeclaringModule)) {
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs]: Why not the `stubMemberType.IsBuiltin`?

File: src/Analysis/Ast/Impl/Caching/StubCache.cs [hunk 5]
```diff
@@ -27,13 +27,15 @@ internal sealed class StubCache : IStubCache {
...
 
-        public StubCache(IServiceContainer services, string cacheRootFolder = null) {
+        public StubCache(IServiceContainer services) {
             _fs = services.GetService<IFileSystem>();
             _log = services.GetService<ILogger>();
+            _cfs = services.GetService<ICacheFolderService>();
 
-            cacheRootFolder = cacheRootFolder ?? CacheFolders.GetCacheFolder(services);
-            StubCacheFolder = Path.Combine(cacheRootFolder, $"stubs.v{_stubCacheFormatVersion}");
+            _cfs = services.GetService<ICacheFolderService>();
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Caching/StubCache.cs]: `ICacheFolderService` is accessed twice.

File: src/Analysis/Ast/Impl/Modules/PythonModule.Analysis.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+using Microsoft.Python.Analysis.Values;
+using Microsoft.Python.Core;
+
+namespace Microsoft.Python.Analysis.Modules {
+    /// <summary>
+    /// Primary base for all modules and user documents. Provides access
+    /// to AST and the module analysis.
+    /// </summary>
+    [DebuggerDisplay("{Name} : {ModuleType}")]
+    internal partial class PythonModule {
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Modules/PythonModule.Analysis.cs]: Does this file has any new code?

File: src/Analysis/Ast/Impl/Modules/Resolution/MainModuleResolution.cs [hunk 7]
```diff
@@ -99,7 +99,12 @@ public MainModuleResolution(string root, IServiceContainer services)
                     return module;
                 }
             }
-            
+
+            var dbs = _services.GetService<IModuleDatabaseService>();
+            if (dbs != null && dbs.TryCreateModule(name, moduleImport.ModulePath, out var m) != ModuleStorageState.DoesNotExist && m != null) {
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Modules/Resolution/MainModuleResolution.cs]: So we are still using this module for `partial`?

File: src/Analysis/Ast/Impl/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -15,5 +15,7 @@
...
 [assembly: InternalsVisibleTo("Microsoft.Python.Analysis.Tests, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
+[assembly: InternalsVisibleTo("Microsoft.Python.Analysis.Caching.Tests, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Properties/AssemblyInfo.cs]: Do we really need these internals?

File: src/Analysis/Ast/Impl/Specializations/Typing/Types/AnyType.cs [hunk 9]
```diff
@@ -24,6 +24,7 @@ internal sealed class AnyType : LocatedMember, IPythonType {
 
         public override PythonMemberType MemberType => PythonMemberType.Class;
         public string Name => "Any";
+        public string QualifiedName => $"{DeclaringModule.Name}:{Name}";
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Specializations/Typing/Types/AnyType.cs]: Can we make it an extension method to `IPythonType`?

File: src/Analysis/Ast/Impl/Types/ParameterInfo.cs [hunk 10]
```diff
@@ -29,7 +29,7 @@ public ParameterInfo(PythonAst ast, Parameter p, IPythonType type, IMember defau
         }
 
         public ParameterInfo(string name, IPythonType type, ParameterKind? kind, IMember defaultValue) {
-            Name = name ?? throw new ArgumentNullException(nameof(name));
+            Name = name ?? "*"; // ?? throw new ArgumentNullException(nameof(name));
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Types/ParameterInfo.cs]: Is it temporary?

File: src/Analysis/Ast/Impl/Types/PythonClassType.cs [hunk 11]
```diff
@@ -196,7 +197,7 @@ internal PythonClassType(string name, Location location)
                 // During processing of builtins module some types may not be available yet.
                 // Specialization will attach proper base at the end.
                 Debug.Assert(!objectType.IsUnknown());
-                if (!bases.Any(b => objectType.Equals(b))) {
+                if (!bases.Any(b => b.Name == "object" && b.DeclaringModule.Name == @"builtins")) {
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Types/PythonClassType.cs]: Why previous comparison didn't work? We can have two instances of `objectType`?

File: src/Analysis/Ast/Impl/Types/PythonUnionType.cs [hunk 12]
```diff
@@ -42,15 +44,11 @@ internal sealed class PythonUnionType : LocatedMember, IPythonUnionType {
         #region IPythonType
 
         public string Name {
-            get {
-                lock (_lock) {
-                    return CodeFormatter.FormatSequence("Union", '[', _types.ToArray());
-                }
-            }
+            get { lock (_lock) { return CodeFormatter.FormatSequence("Union", '[', _types.ToArray()); } }
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Types/PythonUnionType.cs]: I can't say that it is easier to read one line.

File: src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs [hunk 13]
```diff
@@ -26,37 +26,37 @@ namespace Microsoft.Python.Analysis.Values.Collections {
...
-            _interpreter = interpreter;
+            _declaringModule = declaringModule;
         }
 
-        public PythonDictionary(IPythonInterpreter interpreter, IReadOnlyDictionary<IMember, IMember> contents, bool exact = false) :
-            this(new PythonDictionaryType(interpreter), contents, exact: exact) {
-            _interpreter = interpreter;
+        public PythonDictionary(IPythonModule declaringModule, IReadOnlyDictionary<IMember, IMember> contents, bool exact = false) :
+            this(new PythonDictionaryType(declaringModule), contents, exact: exact) {
+            _declaringModule = declaringModule;
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs]: I don't see where we use `_declaringModule` directly. Looks like it can be still `_interpreter = declaringModule.Interpreter`

File: src/Analysis/Ast/Test/Microsoft.Python.Analysis.Tests.csproj [hunk 14]
```diff
@@ -42,6 +42,9 @@
     <ProjectReference Include="..\..\Core\Impl\Microsoft.Python.Analysis.Core.csproj" />
     <ProjectReference Include="..\Impl\Microsoft.Python.Analysis.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Properties\" />
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Test/Microsoft.Python.Analysis.Tests.csproj]: Do we need this?

File: src/PLS.sln [hunk 15]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.27929.0
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.29003.217
```
Comment (AlexanderSher) [inline src/PLS.sln]: Are we switching to VS19?

File: src/LanguageServer/Impl/Implementation/Server.cs [hunk 16]
```diff
@@ -142,6 +143,10 @@ public sealed partial class Server : IDisposable {
                 TypeshedPath = @params.initializationOptions.typeStubSearchPaths.FirstOrDefault()
             };
 
+            if (@params.initializationOptions.enableAnalysCache != false) {
+                //_services.AddService(new ModuleDatabase(_services));
```
Comment (AlexanderSher) [inline src/LanguageServer/Impl/Implementation/Server.cs]: I think we can uncomment this in the branch

File: src/LanguageServer/Impl/Documentation/DocstringConverter.cs [hunk 17]
```diff
@@ -87,7 +87,7 @@ private int NextBlockIndent
         private string CurrentLineWithinBlock => CurrentLine.Substring(_blockIndent);
 
         private DocstringConverter(string input) {
-            _builder = new StringBuilder(input.Length);
+            _builder = new StringBuilder(input?.Length ?? 0);
```
Comment (AlexanderSher) [inline src/LanguageServer/Impl/Documentation/DocstringConverter.cs]: This constructor is called only from:
```cs
        public static string ToMarkdown(string docstring) {
            if (string.IsNullOrWhiteSpace(docstring)) {
                return string.Empty;
            }

            return new DocstringConverter(docstring).Convert();
        }
```
so it shouldn't be null

File: src/Core/Impl/Extensions/StringExtensions.cs [hunk 18]
```diff
@@ -302,5 +302,15 @@ public static bool TryGetNextNonEmptySpan(this string s, char separator, ref (in
             lineEnding = lineEnding ?? Environment.NewLine;
             return string.Join(lineEnding, s.SplitLines());
         }
+
+        public static int GetStableHash(this string s) {
```
Comment (AlexanderSher) [inline src/Core/Impl/Extensions/StringExtensions.cs]: Why not use SHA256?

File: src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    <LangVersion>7.2</LangVersion>
+  </PropertyGroup>
+  <Import Project="..\..\..\build\NetStandard.settings" />
+  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+  <ItemGroup>
+    <None Remove="app.config" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.0.0" />
+    <PackageReference Include="Microsoft.Win32.Registry" Version="4.5.0" />
```
Comment (AlexanderSher) [inline src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj]: Registry?

File: src/LanguageServer/Impl/Documentation/DocstringConverter.cs (hunk 17 above)
Comment (jakebailey) [reply src/LanguageServer/Impl/Documentation/DocstringConverter.cs]: Probably just missing a merge of #1193.

File: src/Caching/Impl/TypeNames.cs [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+
+        /// <summary>
+        /// Splits qualified type name in form of i:A(3.6).B.C into parts. as well as determines if
+        /// qualified name designates instance (prefixed with 'i:').
+        /// </summary>
+        /// <param name="qualifiedName">Qualified name to split. May include instance prefix.</param>
+        /// <param name="moduleName">Module name.</param>
+        /// <param name="memberNames">Module member names such as 'A', 'B', 'C' from module:A.B.C.</param>
+        /// <param name="isInstance">If true, the qualified name describes instance of a type.</param>
+        public static bool DeconstructQualifiedName(string qualifiedName, out string moduleName, out IReadOnlyList<string> memberNames, out bool isInstance) {
```
Comment (AlexanderSher) [inline src/Caching/Impl/TypeNames.cs]: Should we have construction of qualified names here as well?

File: src/Caching/Impl/PythonDbModule.cs [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using Microsoft.Python.Analysis.Caching.Models;
+using Microsoft.Python.Analysis.Modules;
+using Microsoft.Python.Core;
+
+namespace Microsoft.Python.Analysis.Caching {
+    internal sealed class PythonDbModule : SpecializedModule {
```
Comment (AlexanderSher) [inline src/Caching/Impl/PythonDbModule.cs]: Why we are deriving from `SpecializedModule` rather than from `PythonModule`?

File: src/Caching/Impl/ModuleUniqueId.cs [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+using System.Linq;
+using System.Security.Cryptography;
+using Microsoft.Python.Analysis.Core.Interpreter;
+using Microsoft.Python.Analysis.Modules;
+using Microsoft.Python.Analysis.Types;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.IO;
+
+namespace Microsoft.Python.Analysis.Caching {
+    internal static class ModuleUniqueId {
```
Comment (AlexanderSher) [inline src/Caching/Impl/ModuleUniqueId.cs]: Shouldn't it be part of `IModuleDatabaseService`?

File: src/Caching/Impl/ModuleDatabase.cs [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Locates database file based on module information. Module is identified
+        /// by name, version, current Python interpreter version and/or hash of the
+        /// module content (typically file sizes).
+        /// </summary>
+        private string FindDatabaseFile(string moduleName, string filePath) {
+            var interpreter = _services.GetService<IPythonInterpreter>();
```
Comment (AlexanderSher) [inline src/Caching/Impl/ModuleDatabase.cs]: So now we always have one interpreter per LS, right?

Comment (MikhailArkhipov) [reply src/Caching/Impl/ModuleDatabase.cs]: We always actually had... There was notion to have multiple service containers for multi-root workspaces but how actually this is handled is that extension runs multiple language servers, one per root.

File: src/Caching/Impl/ModuleUniqueId.cs (hunk 22 above)
Comment (MikhailArkhipov) [reply src/Caching/Impl/ModuleUniqueId.cs]: Maybe... however, service is for analysis needs and unique id is not used anywhere outside of the caching assembly.

File: src/Caching/Impl/PythonDbModule.cs (hunk 21 above)
Comment (MikhailArkhipov) [reply src/Caching/Impl/PythonDbModule.cs]: Just seemed to be a reasonable choice since specialized is already handled in analysis as something that does not have content or AST and specialized functions are not expected to have bodies or otherwise evaluated.

File: src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj (hunk 19 above)
Comment (MikhailArkhipov) [reply src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj]: We actually have it in all test assemblies. Might be that test framework is bringing it in?

File: src/Caching/Impl/PythonDbModule.cs (hunk 21 above)
Comment (AlexanderSher) [reply src/Caching/Impl/PythonDbModule.cs]: Got it

File: src/Caching/Impl/ModuleDatabase.cs (hunk 23 above)
Comment (AlexanderSher) [reply src/Caching/Impl/ModuleDatabase.cs]: Ok

File: src/Caching/Impl/Models/FunctionAttributes.cs [hunk 24]
```diff
@@ -15,16 +15,12 @@
 
 using System;
 
-namespace Microsoft.Python.Analysis.Types {
-    public interface IPythonFile {
```
Comment (AlexanderSher) [inline src/Caching/Impl/Models/FunctionAttributes.cs]: So this interface isn't needed any longer?

File: src/Caching/Impl/Models/ClassModel.cs [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+            // Skip certain members in order to avoid infinite recursion.
+            foreach (var name in cls.GetMemberNames().Except(new[] { "__base__", "__bases__", "__class__", "mro" })) {
+                var m = cls.GetMember(name);
+
+                // Only take members from this class, skip members from bases.
+                if (m is IPythonClassMember cm && !cls.Equals(cm.DeclaringType)) {
+                    continue;
+                }
+
+                if (!_processing.Push(m)) {
```
Comment (AlexanderSher) [inline src/Caching/Impl/Models/ClassModel.cs]: Is it always single-threaded?

File: src/Core/Impl/Extensions/StringExtensions.cs (hunk 18 above)
Comment (MikhailArkhipov) [reply src/Core/Impl/Extensions/StringExtensions.cs]: Need an int for Id field required by BSON/LiteDb. SHA256 yields `byte[]`, no? Would need to xor then anyway.

File: src/LanguageServer/Impl/Implementation/Server.cs (hunk 16 above)
Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Implementation/Server.cs]: Done

File: src/PLS.sln (hunk 15 above)
Comment (MikhailArkhipov) [reply src/PLS.sln]: Everyone is on it AFAIK... I am on dogfood, works OK, some stuff is better/faster.

File: src/Analysis/Ast/Test/Microsoft.Python.Analysis.Tests.csproj (hunk 14 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Test/Microsoft.Python.Analysis.Tests.csproj]: There is `AssemblyInfo` there with `InternalsVisibleTo("Microsoft.Python.Analysis.Caching.Tests)` for fluent things.

File: src/Core/Impl/Extensions/StringExtensions.cs (hunk 18 above)
Comment (AlexanderSher) [reply src/Core/Impl/Extensions/StringExtensions.cs]: Ah, right. I thought it should be string.

File: src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs (hunk 13 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Values/Collections/PythonDictionary.cs]: OK

File: src/Analysis/Ast/Impl/Types/PythonUnionType.cs (hunk 12 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Types/PythonUnionType.cs]: OK

File: src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs (hunk 2 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs]: No, should never be null. And modules always have names.

File: src/Analysis/Ast/Impl/Caching/StubCache.cs (hunk 5 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Caching/StubCache.cs]: Fixed

File: src/Analysis/Ast/Impl/Modules/PythonModule.Analysis.cs (hunk 6 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Modules/PythonModule.Analysis.cs]: No, just split

File: src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs (hunk 2 above)
Comment (AlexanderSher) [reply src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs]: Can we get here from unsaved file?

File: src/Analysis/Ast/Impl/Properties/AssemblyInfo.cs (hunk 8 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Properties/AssemblyInfo.cs]: Fluent extensions come handy

File: src/Analysis/Ast/Impl/Specializations/Typing/Types/AnyType.cs (hunk 9 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Specializations/Typing/Types/AnyType.cs]: Yes

File: src/Caching/Impl/Models/FunctionAttributes.cs (hunk 24 above)
Comment (MikhailArkhipov) [reply src/Caching/Impl/Models/FunctionAttributes.cs]: It was never used anywhere, doesn't seem to provide any value.

Comment (AlexanderSher) [reply src/Caching/Impl/Models/FunctionAttributes.cs]: Ok

File: src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj (hunk 19 above)
Comment (AlexanderSher) [reply src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj]: Maybe a copy/paste error.

File: src/PLS.sln (hunk 15 above)
Comment (AlexanderSher) [reply src/PLS.sln]: Ok, I was still on VS15

File: src/Analysis/Ast/Test/Microsoft.Python.Analysis.Tests.csproj (hunk 14 above)
Comment (AlexanderSher) [reply src/Analysis/Ast/Test/Microsoft.Python.Analysis.Tests.csproj]: Ok

File: src/Analysis/Ast/Impl/Properties/AssemblyInfo.cs (hunk 8 above)
Comment (AlexanderSher) [reply src/Analysis/Ast/Impl/Properties/AssemblyInfo.cs]: Ok

File: src/Analysis/Ast/Impl/Modules/PythonModule.Analysis.cs (hunk 6 above)
Comment (AlexanderSher) [reply src/Analysis/Ast/Impl/Modules/PythonModule.Analysis.cs]: Ok

File: src/LanguageServer/Impl/Documentation/DocstringConverter.cs (hunk 17 above)
Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Documentation/DocstringConverter.cs]: Yes

File: src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj (hunk 19 above)
Comment (MikhailArkhipov) [reply src/Caching/Test/Microsoft.Python.Analysis.Caching.Tests.csproj]: Removed.

File: src/Analysis/Ast/Impl/Types/Collections/PythonCollectionType.cs (hunk 1 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Types/Collections/PythonCollectionType.cs]: Yes and yes. Collections were assumed to be builtins like `list`, but `List[str]` is actually `typing` and restoration from the db didn't work since we ended up with `builtins:List[str]` that does not make sense.

File: src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Collections.cs (hunk 3 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Analyzer/Evaluation/ExpressionEval.Collections.cs]: `dict` technically should be `builtins:dict` since it is where the type is declared.

File: src/Analysis/Ast/Impl/Types/ParameterInfo.cs (hunk 10 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Types/ParameterInfo.cs]: The problem is that sometimes argument does not have name. This happens to `*` arguments since parser strips the `*`. So I'd like to have at least something.

File: src/Analysis/Ast/Impl/Modules/Resolution/MainModuleResolution.cs (hunk 7 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Modules/Resolution/MainModuleResolution.cs]: `Partial` may or may not happen. The idea was to use Python process to obtain initial info on the module members quickly rather than wait for analysis. Consider `plotly`, for example. So we get quick members from Python, then, when/if analysis completes (user may as well close the app before it finishes), write the actual info. However, I am not 100% confident we'll do it, but we might still.

File: src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs (hunk 2 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Specializations/Typing/Values/TypingType.cs]: Yes but it is specialized typing so it is always actually `typing`.

File: src/Caching/Impl/Models/ClassModel.cs (hunk 25 above)
Comment (MikhailArkhipov) [reply src/Caching/Impl/Models/ClassModel.cs]: Should be as we should be writing as part of the analysis so as long as analysis pass is single and there is only one event leading to the db writing.

File: src/Caching/Impl/TypeNames.cs (hunk 20 above)
Comment (MikhailArkhipov) [reply src/Caching/Impl/TypeNames.cs]: We sort of do, but it is different for modules, types and class/functions.

File: src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs (hunk 4 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs]: Yep

File: src/Analysis/Ast/Impl/Types/PythonClassType.cs (hunk 11 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Types/PythonClassType.cs]: This probably can be reverted. The problem was when real analysis had one builtins module and restored code had another one.

File: src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs (hunk 4 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs]: Actually no. We have a bit of a mess with what is `IsBuiltin` on the type. Ex `PythonModule` has `IsBuiltin => true` always. So does `PythonVariableModule`. Or `Union` when it has all types as builtin types... Modules assume they are builtin if `IsBuiltin => IsCompiled && ModulePath == null`.

This specific check is if member actually belongs to the `builtins` module. We should go clean this up, but it may be some other time.

File: src/Caching/Impl/TypeNames.cs (hunk 20 above)
Comment (AlexanderSher) [reply src/Caching/Impl/TypeNames.cs]: Ok
